[parcel 2, regression] parcel optimizes away parentheses and makes valid code invalid / makes terser fail

# 🐛 bug report

I am lacking words for this issue as I don't know what the features are called. But basically code where the result of an assignment to a variable is `await`ed is correctly built in `parcel@2.0.0-nightly.511` but breaks shortly thereafter and is broken in the newest version.

## 🎛 Configuration (.babelrc, package.json, cli command)

Please checked attached .zip file, run `npm install` and then `npm run build`


## 🤔 Expected Behavior

the code should transpile to this or similar, like it does in version 511:

```
! function() {
  var t = {
    promise: () => {}
  };
  console.log(async function() {
    return await (t.promise = (await fetch("/")).text())
  }(), t)
}();
```

## 😯 Current Behavior

Parcel removes parantheses around the definition of the variable, breaking the code:
```
daniel@iMac:/tmp/parcel-pleeeaase-wtf$ npm run build

> build
> parcel build src/main.ts --log-level verbose --detailed-report 100 --target modern

🚨 Build failed.
@parcel/optimizer-terser: Invalid assignment
/private/tmp/parcel-pleeeaase-wtf/src/main.ts:3:31
  2 | async function main(){
> 3 |   return await (obj.promise = (await fetch("/")).text())
>   |                               ^ Invalid assignment
  4 | }
  5 | 
It's likely that Terser doesn't support this syntax yet.
@parcel/optimizer-terser: Invalid assignment
  5 |   async function $244ca7e764f37acfc6c3db0b97d89bba$var$main() {
> 6 |     return await $244ca7e764f37acfc6c3db0b97d89bba$var$obj.promise = (await fetch("/")).text();
>   |                                                                   ^ Invalid assignment
  7 |   }
  8 |   console.log($244ca7e764f37acfc6c3db0b97d89bba$var$main(), $244ca7e764f37acfc6c3db0b97d89bba$var$obj);
It's likely that Terser doesn't support this syntax yet.
npm ERR! code 1
npm ERR! path /private/tmp/parcel-pleeeaase-wtf
npm ERR! command failed
npm ERR! command sh -c parcel build src/main.ts --log-level verbose --detailed-report 100 --target modern

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/daniel/.npm/_logs/2021-02-03T10_59_06_681Z-debug.log
```

When running parcel with `--no-optimize` it successfully builds to this:

```
(function () {
  var $244ca7e764f37acfc6c3db0b97d89bba$var$obj = {
    promise: () => {}
  };
  async function $244ca7e764f37acfc6c3db0b97d89bba$var$main() {
    return await $244ca7e764f37acfc6c3db0b97d89bba$var$obj.promise = (await fetch("/")).text();
  }
  console.log($244ca7e764f37acfc6c3db0b97d89bba$var$main(), $244ca7e764f37acfc6c3db0b97d89bba$var$obj);
})();
```

But that code is illegal and throws `Uncaught SyntaxError: invalid assignment left-hand side` when being executed.

## 💁 Possible Solution

Revert back to what the old version did which is preserving the parantheses and not breaking the code.

## 🔦 Context

I moved all my css imports to use `bundle-text` after I got to know [this workaround which makes the css minified](https://github.com/parcel-bundler/parcel/issues/5755#issuecomment-772085081). When building all my scripts again to check the transpiled code for unminified css one project failed to build. I investigated and opened this bug report.


## 💻 Code Sample

[parcel-parantheses-bug.zip](https://github.com/parcel-bundler/parcel/files/5917942/parcel-parantheses-bug.zip)

This is the line it can't deal with: `return await (obj.promise = (await fetch("/")).text())`


## 🌍 Your Environment

<!--- Include as many relevant details about the environment you experienced the bug in -->

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.0.0-nightly.566+dba96de0 |
| Node             | v15.6.0 |
| npm/Yarn         | 7.5.2 |
| Operating System | macOS High Sierra 10.13.6 |

--- a/packages/shared/babel-ast-utils/package.json
+++ b/packages/shared/babel-ast-utils/package.json
@@ -23,6 +23,6 @@
     "@babel/parser": "^7.0.0",
     "@parcel/source-map": "2.0.0-alpha.4.19",
     "@parcel/utils": "2.0.0-beta.1",
-    "astring": "^1.6.0"
+    "astring": "^1.6.2"
   }
 }

--- a/packages/shared/babel-ast-utils/src/generator.js
+++ b/packages/shared/babel-ast-utils/src/generator.js
@@ -1,4 +1,4 @@
-import {baseGenerator, EXPRESSIONS_PRECEDENCE} from 'astring';
+import {GENERATOR, EXPRESSIONS_PRECEDENCE} from 'astring';
 
 export const expressionsPrecedence = {
   ...EXPRESSIONS_PRECEDENCE,
@@ -18,7 +18,7 @@ export const expressionsPrecedence = {
 // Convert Babel's AST format to ESTree on the fly.
 // See https://babeljs.io/docs/en/babel-parser#output
 export const generator = {
-  ...baseGenerator,
+  ...GENERATOR,
   Program(node, state) {
     // Monkeypatch state to fix sourcemap filenames.
     let map = state.map;
@@ -34,11 +34,11 @@ export const generator = {
     }
 
     handleDirectives(node);
-    baseGenerator.Program.call(this, node, state);
+    GENERATOR.Program.call(this, node, state);
   },
   BlockStatement(node, state) {
     handleDirectives(node);
-    baseGenerator.BlockStatement.call(this, node, state);
+    GENERATOR.BlockStatement.call(this, node, state);
   },
   NumericLiteral(node, state) {
     node.type = 'Literal';
@@ -68,7 +68,7 @@ export const generator = {
       pattern: node.pattern,
       flags: node.flags,
     };
-    baseGenerator.Literal(node, state);
+    GENERATOR.Literal(node, state);
   },
   BigIntLiteral(node, state) {
     node.type = 'Literal';
@@ -84,7 +84,7 @@ export const generator = {
       // Make sure they don't start with "O"
       node.body.type = '_' + node.body.type;
     }
-    baseGenerator.ArrowFunctionExpression.call(this, node, state);
+    GENERATOR.ArrowFunctionExpression.call(this, node, state);
   },
   ObjectProperty(node, state) {
     node.type = 'Property';
@@ -195,11 +195,11 @@ export const generator = {
   OptionalMemberExpression(node, state) {
     node.optional = true;
     node.type = 'MemberExpression';
-    baseGenerator.MemberExpression.call(this, node, state);
+    GENERATOR.MemberExpression.call(this, node, state);
   },
   MemberExpression(node, state) {
     if (node.optional) node.optional = false;
-    baseGenerator.MemberExpression.call(this, node, state);
+    GENERATOR.MemberExpression.call(this, node, state);
   },
   _OptionalCallExpression(node, state) {
     this.OptionalCallExpression(node, state, true);
@@ -207,11 +207,11 @@ export const generator = {
   OptionalCallExpression(node, state) {
     node.optional = true;
     node.type = 'CallExpression';
-    baseGenerator.CallExpression.call(this, node, state);
+    GENERATOR.CallExpression.call(this, node, state);
   },
   CallExpression(node, state) {
     if (node.optional) node.optional = false;
-    baseGenerator.CallExpression.call(this, node, state);
+    GENERATOR.CallExpression.call(this, node, state);
   },
   ExportNamedDeclaration(node, state) {
     if (node.source) {
@@ -232,7 +232,7 @@ export const generator = {
       }
     }
 
-    baseGenerator[node.type].call(this, node, state);
+    GENERATOR[node.type].call(this, node, state);
   },
   ReturnStatement(node, state) {
     // Add parentheses if there are leading comments
@@ -246,7 +246,7 @@ export const generator = {
       state.write(state.lineEnd);
       state.write(indent + ');');
     } else {
-      baseGenerator.ReturnStatement.call(this, node, state);
+      GENERATOR.ReturnStatement.call(this, node, state);
     }
   },
 };

--- a/packages/shared/babel-ast-utils/test/fixtures/async.js
+++ b/packages/shared/babel-ast-utils/test/fixtures/async.js
@@ -9,3 +9,9 @@ async function j() {
   await g().a;
   return await f();
 }
+async function k() {
+  return await (obj = Promise.resolve(1));
+}
+async function l() {
+  await (() => 1);
+}

--- a/packages/shared/babel-ast-utils/test/fixtures/await.js
+++ b/packages/shared/babel-ast-utils/test/fixtures/await.js
@@ -1,3 +1,0 @@
-async function f() {
-  await (() => 1);
-}

